### DISCUSSION TITLE
Check and install libraries in OpenModelica through command lines?

### DISCUSSION THREAD

**Discussion by ejperezHI** [OpenModelica/OpenModelica#12083] [Q&A] [answered]
Hello everyone

Is there any way of installing (and checking) libraries in my OpenModelica via commands?

I guess it can be done in two ways (correct me if I am wrong):

- Via OMShell

- Via Test commands like omc.sendExpression() explained [here](https://openmodelica.org/doc/OpenModelicaUsersGuide/v1.12.0/ompython.html#ompython)

However, how can I install OMShell? And how can I install libraries thorugh "send.Expression" commands? I cannot find the correct one

Note: The point afterall is being able to check libraries installed and install new ones WITHOUT the gui/api of OMEdit.

Thanks in advance!

**Comment by casella** [comment]
On Windows, OMShell is installed alongside all other OM tools. Are you on Windows or Linux?

**Reply by ejperezHI** [reply]
Hello and thanks for the response,

I have OM in a docker with a Linux machine

**Reply by AnHeuermann** [reply]
To install OMShell on Debian/Ubuntu use `apt` to install package `omshell`. Follow the same procedure from https://openmodelica.org/download/download-linux/ to get OpenModelica packages.

**Comment by AnHeuermann** [comment]
You can automate writing a short installation mos script and run that with `omc`.

The generated script could look something like

*installLibs.mos*
```modelica
updatePackageIndex(); getErrorString();

if not installPackage(Modelica, "4.0.0+maint.om", exactMatch=true) then
  print("Failed to install Modelica 4.0.0+maint.om\n");
  print(getErrorString());
  exit(1);
else
  print("Installed: Modelica\n");
end if;
```

```bash
omc installLibs.mos
```

Here is a similar script that installs libraries for testing of OpenModelica: https://github.com/OpenModelica/OpenModelica/blob/master/libraries/install-index.mos

The same commands can be used from OMShell with sendExpression, just make sure to escape the escape character `\`.

**Reply by adeas31** [reply]
[getAvailableLibraries()](https://build.openmodelica.org/Documentation/OpenModelica.Scripting.getAvailableLibraries.html) tells you which ones are already installed.

**Reply by ejperezHI** [reply]
Thanks for your replies!

@AnHeuermann 
I am checking the link you just posted of github, trying to emulate what he writes there, and taking into account the sintax of the command "installPackage" written [here](https://build.openmodelica.org/Documentation/OpenModelica.Scripting.installPackage.html)

And I get this error when trying to execute that command:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/499809d8-b09e-4f61-90a5-1e242a56913e)


@adeas31 
Thanks for that, I already know that command, and it returns me "()", telling me that I do not have any libraries installed. That is the issue, I want to install them (at least the standard one) but not using the gui/api, but only through commands or cmd.

**Reply by AnHeuermann** [reply, marked as answer]
Escape `"` with `\"` in the middle of the string.
```
omc.sendExpression("installPackage(Modelica, \"4.0.0+maint.om\", exactMatch=true")
```

**Reply by adeas31** [reply]
It should be `omc.sendExpression("installPackage(Modelica, \"4.0.0+maint.om\", exactMatch=true)")`

**Reply by casella** [reply]
For the record, I know nobody ever reads the instruction manual, but there is a section of the User's Guide on this topic, see [here](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/packagemanager.html#using-the-package-manager-from-the-interactive-environment).

**Reply by ejperezHI** [reply]
That was it big thanks all of you!
I do read the User Guide, but didn't find that thing particularly
